presto-main-base/src/test/java/com/facebook/presto/sql/expressions/AbstractTestExpressionInterpreter.java

```diff
@@ -0,0 +1,1697 @@
+    {
+        assertThrows(PrestoException.class, () -> optimize("MAP(ARRAY [1, 2], ARRAY [3, 4])[-1]"));
+    }
+
+    @Test
+    public void testMapSubscriptConstantIndexes()
+    {
+        optimize("MAP(ARRAY [1, 2], ARRAY [3, 4])[1]");
+        optimize("MAP(ARRAY [BIGINT '1', 2], ARRAY [3, 4])[1]");
+        optimize("MAP(ARRAY [1, 2], ARRAY [3, 4])[2]");
+        optimize("MAP(ARRAY [ARRAY[1,1]], ARRAY['a'])[ARRAY[1,1]]");
+    }
+
+    @Test
+    public void testLiterals()
+    {
+        optimize("date '2013-04-03' + unbound_interval");
+        optimize("time '03:04:05.321' + unbound_interval");
+        optimize("time '03:04:05.321 UTC' + unbound_interval");
+        optimize("timestamp '2013-04-03 03:04:05.321' + unbound_interval");
+        optimize("timestamp '2013-04-03 03:04:05.321 UTC' + unbound_interval");
+
+        optimize("interval '3' day * unbound_long");
+        optimize("interval '3' year * unbound_long");
+    }
+
+    @Test
+    public void testVarbinaryLiteral()
+    {
+        assertEquals(optimize("X'1234'"), Slices.wrappedBuffer((byte) 0x12, (byte) 0x34));
+    }
+
+    public void assertExpressionAndRowExpressionEquals(Object expressionResult, Object rowExpressionResult)
+    {
+        if (rowExpressionResult instanceof RowExpression) {
+            // Cannot be completely evaluated into a constant; compare expressions
+            assertTrue(expressionResult instanceof Expression);
+
+            // It is tricky to check the equivalence of an expression and a row expression.
+            // We rely on the optimized translator to fill the gap.
+            RowExpression translated = TRANSLATOR.translateAndOptimize((Expression) expressionResult, SYMBOL_TYPES);
+            assertRowExpressionEvaluationEquals(translated, rowExpressionResult);
+        }
+        else {
+            // We have constants; directly compare
+            assertRowExpressionEvaluationEquals(expressionResult, rowExpressionResult);
+        }
+    }
+
+    public static void assertRowExpressionEvaluationEquals(RowExpression left, RowExpression right)
+    {
+        assertTrue(left instanceof RowExpression);
+        // assertEquals(((RowExpression) left).getType(), ((RowExpression) right).getType());
+        if (left instanceof ConstantExpression) {
+            if (isRemovableCast(right)) {
+                assertRowExpressionEvaluationEquals(left, ((CallExpression) right).getArguments().get(0));
+                return;
+            }
+            assertTrue(right instanceof ConstantExpression);
+            assertRowExpressionEvaluationEquals(((ConstantExpression) left).getValue(), ((ConstantExpression) left).getValue());
+        }
+        else if (left instanceof InputReferenceExpression || left instanceof VariableReferenceExpression) {
+            assertEquals(left, right);
+        }
+        else if (left instanceof CallExpression) {
+            assertTrue(right instanceof CallExpression);
```

Comment:
Abstract function for this logic. Same for SpecialFormExpression and LambdaExpression.